fennel/datasets/datasets.py

```diff
@@ -410,45 +453,100 @@ def dsschema(self):
         )
 
 
+class UDFType(str, Enum):
+    expr = "expr"
+    python = "python"
+
+
 class Assign(_Node):
     def __init__(
-        self, node: _Node, column: str, output_type: Type, func: Callable
+        self,
+        node: _Node,
+        column: Optional[str],
+        output_type: Optional[Type],
+        func: Optional[Callable],
+        **kwargs,
```

Comment:
Maybe do some validation here that if kwargs are given, all of column, output_type, func are None. And if any of them is non-None, all are non-None and kwargs is empty.